### PR TITLE
feat: Secretary, scaffolding for future messages, and cleanup

### PR DESCRIPTION
**Pull request by streeeggs**
# What was added

## Secretary 
- `/secretary` and the subsequent flags `-r`, `-p`, and `-v` should all be respected
- Added a modal when sending the command to make verifying your options (or adding them) easier:
image
_Handles both when the user provides the flags in the message (and subsequently pre-populates them) or leaves them blank_

## `ch-command-modal` mixin
- Should allow for us to reduce some boilerplate if we need to add new modals for additional commands
- Example invocation: `+ch-command-modal("Secretary")`

# What was changed (CyTube core)

## `$("#chatline").on('keydown'...)` in `ui.js`
- Added two new functions `coolholeShouldOverrideMessageSend` and `coolholeMessageOverride`
- Needed these since we had to intercept **outgoing** messages and display a modal if it's a specific CH command

## `addChatMessage` in `utils.js`
- Added two new functions `coolholeShouldShowMessageOutOfBuffer` and `coolholeHandleMessageOutOfBuffer`
- Needed these since we had to intercept **incoming** messages so they don't render inside the message buffer

## `chatTabComplete` in `ui.js`
- Added CH commands to the list of tab complete options

## General use of Coolhole specific `mixin`s
- Trying to standardize all CH mixins with `ch-` prefixed so it'll be easier to identify what's CH specific

# What was changed (Coolhole)

## `Coolpoints` server class
- Added `ActionResult` as a generic return object to hopefully easily describe if an action went through and if it didn't, why not
  - Subsequently spent way too much time trying to get proper JSDoc formatting to show a return object but I don't think it worked...
- Added `isUserEligibleForPoints` to reduce some repetition
-

## `processSpeechMessage` in `coolhole-utils.js`
- Wrote a `parseSecretaryMessage` to parse an incoming message with regex instead since I had to do that to pre-populate the new modal anyway
  - ⚠I'd like to find a way to send this info to the server (and receive it from the server) via a true JS object instead of parsing strings on either end
  - Could use the coolhole meta object some how...

# Tested...
- [X] Gold
- [X] European
- [X] European Gold
- [X] No commands
- [X] Only one of each command
- [X] Only message
- [X] All commands + no message
  - Expected behavior is kinda weird here since I can't reasonably assume someone wouldn't pass `{` and `}` in a message so I just assume they want the flag object as the message
  - This is more so a limitation of the command syntax since the message is technically in the same "grammar" as the flags (a message can be any string and a flag object is a specific string or not there at all)
- [X] Sending multiple secretary commands after one another
- [X] Attempting secretary with insufficient funds 

--- 
I hopefully haven't updated formatting on any pages. I installed prettier and have been tempted to add a config file to the project to keep everything uniform but, I don't want to risk mass updating existing CyTube core files with white space changes...

### CODE REVIEW

File: www/js/coolhole-util.js [hunk 1]
```diff
@@ -4,6 +4,26 @@
...
+ * Globals
+ * ============================================================
+ */
+const synth = window.speechSynthesis;
+let voices = [];
+function populateVoiceList() {
+  voices = synth.getVoices();
+  // If we ever wanna give users a UI for what voices are available, they can select it here
+  // Probably won't translate between different browsers but w/e
+}
```
Comment (streeeggs) [inline www/js/coolhole-util.js]: fix: in chrome, there's an issue where voices aren't populated on load. to fix this, we need to clear and populate the dropdown every time the voices change:
 ```suggestion
const SEC_MSG_INPUT = $("#secretaryOption-message");
const SEC_PITCH_INPUT = $("#secretaryOption-pitch");
const SEC_RATE_INPUT = $("#secretaryOption-rate");
const SEC_VOICE_INPUT = $("#secretaryOption-voice");

const synth = window.speechSynthesis;
let voices = [];
function populateVoiceList() {
  voices = synth.getVoices();
  // If we ever wanna give users a UI for what voices are available, they can select it here
  // Probably won't translate between different browsers but w/e
  
  // Clear out the secretary voices select
  SEC_VOICE_INPUT.empty();
  // Add the voices to the secretary voices select
  voices.forEach((voice) => {
    $("<option/>").val(voice.name).text(voice.name).appendTo(SEC_VOICE_INPUT);
  });
}
```

File: www/js/coolhole-util.js [hunk 2]
```diff
@@ -73,6 +93,334 @@ function isMessageTooOld(msgTime) {
...
+const SEC_PITCH_INPUT = $("#secretaryOption-pitch");
+const SEC_RATE_INPUT = $("#secretaryOption-rate");
+const SEC_VOICE_INPUT = $("#secretaryOption-voice");
+// Populate the voice options
+voices.forEach((voice) => {
+  $("<option/>")
+    .val(voice.name)
+    .text(voice.name)
+    .appendTo(SEC_VOICE_INPUT);
+});
```
Comment (streeeggs) [inline www/js/coolhole-util.js]: see above: https://github.com/streeeggs/coolhole/pull/22/files#r1752682721
```suggestion
```

File: www/js/coolhole-util.js [hunk 3]
```diff
@@ -73,6 +93,334 @@ function isMessageTooOld(msgTime) {
...
+	speechObj.voiceObj = (voice object from voices array)
+	speechObj.message = "this is my message."
+*/
+function processSpeechMessage(chatMessage) {
+  const defaultSpeechObj = {
+    rate: 1.2,
+    pitch: 1,
+    voiceObj: voices.find(
+      (voice) => voice.name === "Microsoft Zira - English (United States)"
+    ),
```
Comment (streeeggs) [inline www/js/coolhole-util.js]: Not everyone loves CBT/Wangblows
```suggestion
    voiceObj: voiceObj: voices.find((voice) => voice.default) ?? voices[0],
```

File: www/js/coolhole-util.js [hunk 4]
```diff
@@ -73,6 +93,334 @@ function isMessageTooOld(msgTime) {
...
+    const speechObj = processSpeechMessage(data.msg);
+
+    // Create the utterance object for the message with modified rate/pitch/voice/message
+    const utterThis = new SpeechSynthesisUtterance(speechObj.message);
+
+    if (speechObj.voiceObj) {
+      utterThis.voice = speechObj.voiceObj;
+      utterThis.rate = speechObj.rate;
+      utterThis.pitch = speechObj.pitch;
+    }
```
Comment (streeeggs) [inline www/js/coolhole-util.js]:
```suggestion
    if (speechObj.voiceObj) 
      utterThis.voice = speechObj.voiceObj;
    if (speechObj.rate)
      utterThis.rate = speechObj.rate;
    if (speechObj.pitch)
      utterThis.pitch = speechObj.pitch;
```

File: src/channel/coolholepoints.js [hunk 5]
```diff
@@ -458,11 +519,12 @@ class Coolpoints extends ChannelModule {
           `Wise spender ${user.getName()} spent ${pointsToSpend} points on ${action}`,
           new ReturnPointData(
             user.getName(),
-            pointsToSpend,
+            -pointsToSpend,
```
Comment (streeeggs) [inline src/channel/coolholepoints.js:522]: found by @esnyder232 , UI was showing "green' when we were sending positive values (duh)

File: src/channel/coolholepoints.js [hunk 6]
```diff
@@ -260,7 +298,7 @@ class Coolpoints extends ChannelModule {
    * @param {Number} points User's coolpoints
    */
   subtract(name, points) {
-    this.set(name, this.get(name)?.points ?? 0 - points);
+    this.set(name, (this.get(name)?.points ?? 0) - points);
```
Comment (streeeggs) [inline src/channel/coolholepoints.js:301]: found by @esnyder232 , learn order of operations

File: www/js/coolhole-util.js (hunk 1 above)
Comment (esnyder232) [reply www/js/coolhole-util.js]: its good

File: www/js/coolhole-util.js (hunk 2 above)
Comment (esnyder232) [reply www/js/coolhole-util.js]: its good

File: www/js/coolhole-util.js (hunk 4 above)
Comment (esnyder232) [reply www/js/coolhole-util.js]: its good

File: www/js/coolhole-util.js (hunk 3 above)
Comment (esnyder232) [reply www/js/coolhole-util.js]: its good

File: www/js/coolhole-util.js [hunk 7]
```diff
@@ -73,6 +105,322 @@ function isMessageTooOld(msgTime) {
...
+	speechObj.rate = 1.2
+	speechObj.pitch = 3.4
+	speechObj.voiceObj = (voice object from voices array)
+	speechObj.message = "this is my message."
+*/
+function processSpeechMessage(chatMessage) {
+  const defaultSpeechObj = {
+    rate: 1.2,
+    pitch: 1,
+    voiceObj: voiceObj: voices.find((voice) => voice.default) ?? voices[0],
```
Comment (esnyder232) [inline www/js/coolhole-util.js]: fix: voiceObj twice causing error in console

**Review by esnyder232 — APPROVED** [review]
Everything looks good!

**Pull request merged**